backends/fs/morphos/morphos-fs.cpp

```diff
@@ -0,0 +1,432 @@
+	// 		nodeExists = true;
+	// 	IDOS->UnLock(pLock);
+	// }
+	//
+	// IDOS->FreeDosObject(DOS_FIB, fib);
+	//
+	// =============================  New code
+	BPTR pLock = Lock(_sPath.c_str(), SHARED_LOCK);
+	if (pLock) {
+		nodeExists = true;
+		UnLock(pLock);
+	}
+
+	//LEAVE();
+	return nodeExists;
+}
+
+AbstractFSNode *MorphOSFilesystemNode::getChild(const Common::String &n) const {
+	//ENTER();
+	if (!_bIsDirectory) {
+		debug(6, "Not a directory");
+		return 0;
+	}
+
+	Common::String newPath(_sPath);
+
+	if (_sPath.lastChar() != '/')
+		newPath += '/';
+
+	newPath += n;
+
+	//LEAVE();
+	return new MorphOSFilesystemNode(newPath);
+}
+
+bool MorphOSFilesystemNode::getChildren(AbstractFSList &myList, ListMode mode, bool hidden) const {
+
+
+	//bool ret = false;
+
+	if (!_bIsValid) {
+		debug(6, "Invalid node");
+		return false; // Empty list
+	}
+
+	if (!_bIsDirectory) {
+		debug(6, "Not a directory");
+		return false; // Empty list
+	}
+
+	if (isRootNode()) {
+		debug(6, "Root node");
+		myList = listVolumes();
+		return true;
+	}
+
+
+	FileInfoBlock *fib = (FileInfoBlock*) AllocDosObject(DOS_FIB, NULL);
+	
+	if (fib == NULL) {
+	 debug(6, "Failed to allocate memory for FileInfoBLock");
+	 return false;
+	}
+	
+	if (Examine(_pFileLock, fib) != DOSFALSE) {
+	
+	  while (ExNext(_pFileLock, fib) != DOSFALSE)
+	  {
+	  
+	  	MorphOSFilesystemNode *entry;
+	  	Common::String full_path = NULL;
+	  	BPTR pLock;
+	  	
+	  	if ((mode == Common::FSNode::kListAll) || 
+	  		(fib->fib_EntryType > 0 && (Common::FSNode::kListDirectoriesOnly == mode)) ||
+	  		(fib->fib_EntryType < 0 && (Common::FSNode::kListFilesOnly == mode )))
+	  		{
+	  		
+	  		full_path = _sPath;
+	  		full_path += fib->fib_FileName;
+	  		pLock = Lock(full_path.c_str(), SHARED_LOCK);
+			if (pLock) {
+					entry = new MorphOSFilesystemNode( pLock, fib->fib_FileName );
+					if (entry) {
+						myList.push_back(entry);
+					}
+
+					UnLock(pLock);
+			}	  
+			
+			}
+			
+		}	
+	  	if (ERROR_NO_MORE_ENTRIES != IoErr() ) {
+			debug(6, "An error occurred during ExamineDir");
+			return false;
+		}
+	}
+	
+	FreeDosObject(DOS_FIB, fib);
+	  
+	return true;
+		
+}
+
+AbstractFSNode *MorphOSFilesystemNode::getParent() const {
+	//ENTER();
+
+	if (isRootNode()) {
+		debug(6, "Root node");
+		//LEAVE();
+		return new MorphOSFilesystemNode(*this);
+	}
+
+	BPTR pLock = _pFileLock;
+
+	if (!_bIsDirectory) {
+		assert(!pLock);
+		pLock = Lock((CONST_STRPTR)_sPath.c_str(), SHARED_LOCK);
+		assert(pLock);
+	}
+
+	MorphOSFilesystemNode *node;
+
+	BPTR parentDir = ParentDir( pLock );
+	if (parentDir) {
+		node = new MorphOSFilesystemNode(parentDir);
+		UnLock(parentDir);
+	} else
+		node = new MorphOSFilesystemNode();
+
+	if (!_bIsDirectory) {
+		UnLock(pLock);
+	}
+
+	//LEAVE();
+
+	return node;
+}
+
+
+AbstractFSList MorphOSFilesystemNode::listVolumes() const {
+
+	AbstractFSList myList;
+	DosList *dosList;
+	const ULONG kLockFlags = LDF_READ | LDF_VOLUMES;
+	char buffer[MAXPATHLEN];
```

Comment:
I'd suggest using `Common::String` here instead of using `strcpy`, `strcat` and `sprintf`.